File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
function getSum(root) {
    if (root == null) {
        return 0;
    }
    return (root.val + getSum(root.left) + getSum(root.right))
}
function countnodes(root){
    if (root==null){
        return 0;
    }
    return (1+ countnodes(root.left)+ countnodes(root.right))
}
function dfs(root, count) {
    if (root == null) {
        return
    }
    // console.log( getSum(root),countnodes(root), root)
    if (root.val === Math.floor(getSum(root)/countnodes(root))) {
        count.val = count.val + 1;
    }

    dfs(root.left, count)

    dfs(root.right, count)
}
var averageOfSubtree = function (root) {


    let count = {
        val: 0
    }
    dfs(root, count);


    return count.val

};

File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    pair<int, int> dfs(TreeNode* root, int& res) {
        if (root == NULL) {
            return {0,0};
        }
        if (root->left == NULL && root->right == NULL) {
            res += 1;
            return {root->val,1};
        }
        pair<int, int> lsum = dfs(root->left, res);
        pair<int, int> rsum = dfs(root->right, res);
        int param = (lsum.first + rsum.first + root->val) /
                    (lsum.second + rsum.second + 1);
        if (param == root->val) {
            res += 1;
        }
        pair<int, int> ans;
        ans.first = lsum.first + rsum.first + root->val;
        ans.second = lsum.second + rsum.second + 1;
        return ans;
    }
    int averageOfSubtree(TreeNode* root) {
        int res = 0;
        pair<int, int> val = dfs(root, res);
        return res;
    }
};